pyomo/core/base/reference.py

```diff
@@ -639,51 +648,82 @@ def Reference(reference, ctype=_NotSpecified):
+        slice_idx = None
+        if reference.is_indexed():
+            index = reference.index_set()
+        else:
+            index = UnindexedComponent_ReferenceSet
+        if ctype is NOTSET:
+            ctype = reference.ctype
         reference = reference[...]
         _data = _ReferenceDict(reference)
-        _iter = iter(reference)
-        slice_idx = []
-        index = None
+        # index and ctype are now set; no need to iterate over the slice
+        _iter = ()
     elif isinstance(reference, ComponentData):
+        slice_idx = None
+        index = UnindexedComponent_ReferenceSet
+        if ctype is NOTSET:
+            ctype = reference.ctype
         reference = IndexedComponent_slice(reference.parent_component())[
             reference.index()]
         _data = _ReferenceDict(reference)
-        _iter = iter(reference)
-        slice_idx = []
-        index = SetOf(UnindexedComponent_set)
+        # index and ctype are now set; no need to iterate over the slice
+        _iter = ()
     elif isinstance(reference, Mapping):
+        slice_idx = None
         _data = _ReferenceDict_mapping(dict(reference))
         _iter = _data.values()
-        slice_idx = None
         index = SetOf(_data)
     elif isinstance(reference, Sequence):
+        slice_idx = None
         _data = _ReferenceDict_mapping(OrderedDict(enumerate(reference)))
         _iter = _data.values()
-        slice_idx = None
         index = OrderedSetOf(_data)
     else:
         raise TypeError(
             "First argument to Reference constructors must be a "
             "component, component slice, Sequence, or Mapping (received %s)"
             % (type(reference).__name__,))
 
-    if ctype is _NotSpecified:
+    if ctype is NOTSET:
         ctypes = set()
     else:
-        # If the caller specified a ctype, then we will prepopulate the
-        # list to improve our chances of avoiding a scan of the entire
-        # Reference (by simulating multiple ctypes having been found, we
-        # can break out as soon as we know that there are not common
-        # subsets).
-        ctypes = set((1,2))
+        if slice_idx is None:
+            # A slice was not provided. We know the ctype and that there
+            # cannot be common subsets (because a slice was not provided).
+            # We don't need to iterate over the data objects at all.
+            # Note that this is redundant for Component and ComponentData
+            # inputs, as _iter was already empty.
+            _iter = ()
+        else:
+            # If the caller specified a ctype, then we will prepopulate
+            # the list to improve our chances of avoiding a scan of the
+            # entire Reference (by simulating multiple ctypes having
+            # been found, we can break out as soon as we know that there
+            # are not common subsets by forcing the len(ctypes) > 1 check
+            # to return True).
+            ctypes = set((1,2))
```

Comment:
Couldn't you just check this condition in the loop below, when you check `len(ctypes)`? This seems confusing.

Reply:
Good call.  Fixed.